File: app/models/employee.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import Date, DateTime, ForeignKey, Index, Numeric, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class Employee(Base):
    __tablename__ = "employees"
    __table_args__ = (
        UniqueConstraint("school_id", "id_card_number", name="uq_employees_school_idcard"),
        Index("ix_employees_school_id", "school_id"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    school_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="RESTRICT"), nullable=False
    )
    first_name: Mapped[str] = mapped_column(String(100), nullable=False)
    middle_name: Mapped[Optional[str]] = mapped_column(String(100))
    last_name: Mapped[str] = mapped_column(String(100), nullable=False)
    birth_date: Mapped[Optional[date]] = mapped_column(Date)
    place_of_birth: Mapped[Optional[str]] = mapped_column(String(255))
    sex: Mapped[Optional[str]] = mapped_column(String(1))
    civil_state: Mapped[Optional[str]] = mapped_column(String(50))
    nationality: Mapped[Optional[str]] = mapped_column(String(100))
    naturality: Mapped[Optional[str]] = mapped_column(String(100))
    height: Mapped[Optional[Decimal]] = mapped_column(Numeric(5, 2))
    profession: Mapped[Optional[str]] = mapped_column(String(255))
    qualifications: Mapped[Optional[str]] = mapped_column(String(255))
    id_card_number: Mapped[Optional[str]] = mapped_column(String(100))
    photo_url: Mapped[Optional[str]] = mapped_column(String(500))
    # Address
    street: Mapped[Optional[str]] = mapped_column(String(255))
    house_number: Mapped[Optional[str]] = mapped_column(String(50))
    building_number: Mapped[Optional[str]] = mapped_column(String(50))
    apt_number: Mapped[Optional[str]] = mapped_column(String(50))
    city: Mapped[Optional[str]] = mapped_column(String(100))
    municipio: Mapped[Optional[str]] = mapped_column(String(100))
    bairro: Mapped[Optional[str]] = mapped_column(String(100))
    # Contacts
    mobile_first: Mapped[Optional[str]] = mapped_column(String(50))
    mobile_second: Mapped[Optional[str]] = mapped_column(String(50))
    email: Mapped[Optional[str]] = mapped_column(String(255))
    # Employment
    employee_type: Mapped[str] = mapped_column(String(50), nullable=False)  # teacher, staff, admin
    position: Mapped[Optional[str]] = mapped_column(String(255))
    title_academic: Mapped[Optional[str]] = mapped_column(String(255))
    social_security: Mapped[Optional[str]] = mapped_column(String(100))
    contract_type: Mapped[Optional[str]] = mapped_column(String(50))  # permanent, temporary, intern
    hire_date: Mapped[Optional[date]] = mapped_column(Date)
    salary: Mapped[Optional[Decimal]] = mapped_column(Numeric(10, 2))
    status: Mapped[str] = mapped_column(String(20), default="active")  # active, inactive, suspended
    privilege: Mapped[Optional[str]] = mapped_column(String(255))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

File: mobile/lib/core/models/employee.dart
class Employee {
  final String id;
  final String schoolId;
  final String firstName;
  final String? middleName;
  final String lastName;
  final String employeeType; // teacher, staff, admin (HR category)
  final List<String> roles; // actual system roles from User.roles
  final String? position;
  final String? cedula;
  final String? taxId;
  final String? socialSecurity;
  final String? contractType;
  final double? salary;
  final String? phone;
  final String? email;
  final String? photoUrl;
  final DateTime? hireDate;
  final bool isActive;
  final String? userId;

  const Employee({
    required this.id,
    required this.schoolId,
    required this.firstName,
    this.middleName,
    required this.lastName,
    required this.employeeType,
    this.roles = const [],
    this.position,
    this.cedula,
    this.taxId,
    this.socialSecurity,
    this.contractType,
    this.salary,
    this.phone,
    this.email,
    this.photoUrl,
    this.hireDate,
    required this.isActive,
    this.userId,
  });

  String get fullName =>
      [firstName, middleName, lastName].whereType<String>().join(' ');

  String get employeeTypeLabel {
    switch (employeeType) {
      case 'teacher':
        return 'Professor / Educador';
      case 'staff':
        return 'Funcionário';
      case 'admin':
        return 'Administrador';
      default:
        return employeeType;
    }
  }

  factory Employee.fromJson(Map<String, dynamic> json) {
    return Employee(
      id: json['id']?.toString() ?? '',
      schoolId: json['school_id']?.toString() ?? '',
      firstName: json['first_name'] as String? ?? '',
      middleName: json['middle_name'] as String?,
      lastName: json['last_name'] as String? ?? '',
      employeeType: json['employee_type'] as String? ?? 'staff',
      roles: (json['roles'] as List?)?.map((e) => e.toString()).toList() ?? [],
      position: json['position'] as String?,
      cedula: json['id_card_number'] as String?,
      taxId: json['tax_id'] as String?,
      socialSecurity: json['social_security'] as String?,
      contractType: json['contract_type'] as String?,
      salary: (json['salary'] as num?)?.toDouble(),
      phone: json['mobile_first'] as String?,
      email: json['email'] as String?,
      photoUrl: json['photo_url'] as String?,
      hireDate: json['hire_date'] != null
          ? DateTime.tryParse(json['hire_date'] as String)
          : null,
      isActive: (json['status'] as String? ?? 'active') == 'active',
      userId: json['user_id']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'school_id': schoolId,
        'first_name': firstName,
        if (middleName != null) 'middle_name': middleName,
        'last_name': lastName,
        'employee_type': employeeType,
        if (position != null) 'position': position,
        if (cedula != null) 'id_card_number': cedula,
        if (taxId != null) 'tax_id': taxId,
        if (socialSecurity != null) 'social_security': socialSecurity,
        if (contractType != null) 'contract_type': contractType,
        if (salary != null) 'salary': salary,
        if (phone != null) 'mobile_first': phone,
        if (email != null) 'email': email,
        if (photoUrl != null) 'photo_url': photoUrl,
        if (hireDate != null)
          'hire_date':
              '${hireDate!.year.toString().padLeft(4, '0')}-${hireDate!.month.toString().padLeft(2, '0')}-${hireDate!.day.toString().padLeft(2, '0')}',
        'status': isActive ? 'active' : 'inactive',
        if (userId != null) 'user_id': userId,
      };
}
